src/cursor.rs

```diff
@@ -1,723 +1,720 @@
+    /// creating new nodes, if possible.
+    fn continue_traverse_to_target(&mut self) {
+        while !self.state.is_leaf() {
+            let (current_node, operation_offset) = core::mem::take(&mut self.state)
+                .into_node_try_mut()
+                .unwrap();
+
+            let next_node = match current_node {
+                TryAsNodeMut::Shared(node) => {
+                    let operated_entry = node.entry(operation_offset);
+                    if !operated_entry.is_node() {
+                        self.reset();
+                        return;
+                    }
 
-            self.move_to(current_node);
-        }
-        self.ref_operated_entry().unwrap()
-    }
+                    self.ancestors.push(node);
 
-    /// Increase the number of heights for XArray to expand its capacity, allowing it to accommodate
-    /// the target index, and returns the height of the final head node.
-    ///
-    /// If the head node of the XArray does not exist, allocate a new head node of appropriate height
-    /// directly. Otherwise, if needed, repeatedly insert new nodes on top of the current head node to
-    /// serve as the new head.
-    fn expand_height(&mut self) -> Height {
-        if self.xa.head().is_null() {
-            let mut head_height = Height::new(1);
-            while self.index > head_height.max_index() {
-                *head_height += 1;
-            }
-            let head = self.alloc_node(head_height, 0);
-            self.xa.set_head(head);
-            return head_height;
-        } else {
-            loop {
-                let head_height = {
-                    if self.is_exclusive {
-                        self.xa.ensure_head_exclusive();
+                    TryAsNodeMut::Shared(operated_entry.as_node_ref().unwrap())
+                }
+                TryAsNodeMut::Exclusive(node) => {
+                    let (operated_entry, dormant_node) = NodeMutRef::new(node, operation_offset);
+                    if !operated_entry.is_node() {
+                        self.reset();
+                        return;
                     }
-                    let head = self.xa.head().as_node().unwrap();
-                    head.height()
-                };
 
-                if head_height.max_index() > self.index {
-                    return head_height;
-                }
+                    self.mut_ancestors.push(dormant_node);
 
-                let new_node_entry = self.alloc_node(Height::new(*head_height + 1), 0);
-                let old_head_entry = self.xa.set_head(new_node_entry);
-                let old_head = old_head_entry.as_node_mut().unwrap();
-                let new_head = self.xa.head().as_node_mut().unwrap();
-                for i in 0..3 {
-                    if !old_head.mark(i).is_clear() {
-                        new_head.set_mark(0, i);
-                    }
+                    operated_entry.try_as_node_mut().unwrap()
```

Comment:
`continue_traverse_to_target` will not trigger cow mechanism and can will be called by `next()`. But I think there might be a possibility where the upper-level node is exclusive, but a portion of the lower-level nodes remain shared. In this case, would the `next` call to this function cause the unwrap here to fail?

Reply:
No, `try_as_node_mut` returns `Option<TryAsNodeMut<'_, I>>`. After `unwrap`, we have a `TryAsNodeMut<'_, I>`, which correctly encodes the two possibilities:
```rust
pub(super) enum TryAsNodeMut<'a, I>
where
    I: ItemEntry,
{
    Shared(&'a XNode<I>),
    Exclusive(&'a mut XNode<I>),
}
```

I think the confusion comes from we use the outer `Option<_>` to encode whether the entry points to a node, and use the inner `TryAsNodeMut<'_, I>` to encode whether the reference to the node is exclusive. Perhaps the APIs can still be improved to avoid this confusion.

Reply:
I get it. It makes sense to me, but the name `TryAsNodeMut` might cause some confusion; it feels like it would generate a type of Option<&mut XNode>.